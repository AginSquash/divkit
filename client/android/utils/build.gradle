apply from: "${project.projectDir}/../div-library.gradle"
apply from: "${project.projectDir}/../div-tests.gradle"
apply from: "${project.projectDir}/../publish-android.gradle"

android {
    namespace 'com.yandex.div.util'
}

dependencies {
    implementation project(path: ':assertion')

    implementation libs.javax
    implementation libs.androidx.annotations
    implementation libs.androidx.collection
    implementation libs.androidx.core
    implementation libs.androidx.viewpager
    implementation(libs.androidx.appcompat)  {
        exclude group: "androidx.fragment", module: "fragment"
    }
}
