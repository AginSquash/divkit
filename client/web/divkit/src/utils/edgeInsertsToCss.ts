import type { EdgeInsets } from '../types/edgeInserts';
import { pxToEm } from './pxToEm';

export function edgeInsertsToCss(edgeInsets: EdgeInsets): string {
    const top = edgeInsets.top || 0;
    const right = edgeInsets.right || 0;
    const bottom = edgeInsets.bottom || 0;
    const left = edgeInsets.left || 0;

    if (top === 0 && right === 0 && bottom === 0 && left === 0) {
        return '';
    }

    return pxToEm(top) +
        ' ' +
        pxToEm(right) +
        ' ' +
        pxToEm(bottom) +
        ' ' +
        pxToEm(left);
}
