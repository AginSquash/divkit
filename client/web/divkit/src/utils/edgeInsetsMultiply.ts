import { MaybeMissing } from '../expressions/json';
import { EdgeInsets } from '../types/edgeInserts';

export function edgeInsertsMultiply(
    edgeInsets: MaybeMissing<EdgeInsets>,
    multiply: number
): MaybeMissing<EdgeInsets> {
    if (!edgeInsets) {
        return {};
    }

    const res: EdgeInsets = {};

    for (const key of ['left', 'top', 'right', 'bottom', 'start', 'end'] as const) {
        const val = edgeInsets[key];
        if (val) {
            res[key] = val * multiply;
        }
    }

    return res;
}
