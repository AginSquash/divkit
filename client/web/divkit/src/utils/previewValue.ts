export function previewValue(value: unknown): string {
    if (typeof value === 'object' && value) {
        if (Array.isArray(value)) {
            return 'array';
        }
        return 'object';
    } else if (value === null) {
        return 'null';
    } else if (value === undefined) {
        return 'undefined';
    }
    return JSON.stringify(value);
}
