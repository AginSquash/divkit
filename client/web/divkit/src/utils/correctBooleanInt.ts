import { previewValue } from './previewValue';
import { LogError, wrapError } from './wrapError';

export function correctBooleanInt(
    val: unknown,
    defaultVal: boolean,
    logError: LogError
): boolean {
    if (val === 1 || val === 0 || val === false || val === true) {
        return Boolean(val);
    }
    if (val !== undefined) {
        logError(wrapError(new Error(`Invalid value: ${previewValue(val)}. Expression<Bool> expected.`)));
    }
    return defaultVal;
}
