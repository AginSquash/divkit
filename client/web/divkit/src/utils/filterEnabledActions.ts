import type { Action, DisappearAction, VisibilityAction } from '../../typings/common';
import type { MaybeMissing } from '../expressions/json';
import { correctBooleanInt } from './correctBooleanInt';
import { LogError } from './wrapError';

export function filterEnabledActions(
    action: MaybeMissing<Action | VisibilityAction | DisappearAction>,
    logError: LogError
): boolean {
    return correctBooleanInt(action.is_enabled, true, logError);
}
