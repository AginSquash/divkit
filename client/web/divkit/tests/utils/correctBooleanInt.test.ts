import { correctBooleanInt } from '../../src/utils/correctBooleanInt';

describe('correctBooleanInt', () => {
    test('simple', () => {
        const logError = jest.fn();

        expect(correctBooleanInt(2, false, logError)).toBe(false);
        expect(correctBooleanInt(2, true, logError)).toBe(true);
        expect(correctBooleanInt(1, false, logError)).toBe(true);
        expect(correctBooleanInt(1, true, logError)).toBe(true);
        expect(correctBooleanInt(0, false, logError)).toBe(false);
        expect(correctBooleanInt(0, true, logError)).toBe(false);
        expect(correctBooleanInt(true, false, logError)).toBe(true);
        expect(correctBooleanInt(true, true, logError)).toBe(true);
        expect(correctBooleanInt(false, false, logError)).toBe(false);
        expect(correctBooleanInt(false, true, logError)).toBe(false);
        expect(correctBooleanInt(undefined, false, logError)).toBe(false);
        expect(correctBooleanInt(undefined, true, logError)).toBe(true);

        expect(logError.mock.calls).toMatchSnapshot();
    });
});
