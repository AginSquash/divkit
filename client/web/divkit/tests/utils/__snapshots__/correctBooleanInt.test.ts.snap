// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctBooleanInt simple 1`] = `
[
  [
    [Error: Invalid value: 2. Expression<Bool> expected.],
  ],
  [
    [Error: Invalid value: 2. Expression<Bool> expected.],
  ],
]
`;
